### PR TITLE
Make setindex! for sparse matrices and vectors not purge stored entries on zero assignment

### PR DESCRIPTION
**Pull request by Sacha0**
This pull request makes `setindex!` for `SparseMatrixCSC`s: (1) not purge allocated entries on zero assignment; and (2) not introduce new allocated entries on zero assignment. See the proposal in  https://github.com/JuliaLang/julia/pull/15568#issuecomment-226645910 / extensive discussion in #15568, #9928, #9906, #6769, #5538, #5454, and probably others. Thanks to @KristofferC for getting the ball rolling with #15568!

~~TODO: Expand tests for new behavior.~~ Now with expanded tests!

### CODE REVIEW

**Comment by Sacha0** [comment]
The `test/show.jl` failure should be fixed. Also added a note to NEWS.md. Thanks!

**Comment by tkelman** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by Sacha0** [comment]
AppVeyor i686 string test failure seems unrelated?

**Comment by tkelman** [comment]
I would assume so, but I've never seen that before
```
From worker 5:	     * string               Test Failed
	From worker 5:	  Expression: [i for i = eachindex("ḟøøƀäṙ")] == [1,4,6,8,10,12]
	From worker 5:	   Evaluated: Array{Dict{VersionNumber,Array{VersionNumber,1}},1}[Dict{VersionNumber,Array{VersionNumber,1}}[#undef],Dict{VersionNumber,Array{VersionNumber,1}}[#undef,#undef,#undef,#undef],Dict{VersionNumber,Array{VersionNumber,1}}[#undef,#undef,#undef,#undef,#undef,#undef],Dict{VersionNumber,Array{VersionNumber,1}}[#undef,#undef,#undef,#undef,#undef,#undef,#undef,#undef],Dict{VersionNumber,Array{VersionNumber,1}}[#undef,#undef,#undef,#undef,#undef,#undef,#undef,#undef,#undef,#undef],Dict{VersionNumber,Array{VersionNumber,1}}[#undef,#undef,#undef,#undef,#undef,#undef,#undef,#undef,#undef,#undef,#undef,#undef]] == [1,4,6,8,10,12]
... snip ...
Exception running test string :
On worker 5:
LoadError: LoadError: There was an error during testing
 in record at .\test.jl:397
 in do_test at .\test.jl:281
 in include_string at .\loading.jl:380
 in include_from_node1 at .\loading.jl:429
 in include_string at .\loading.jl:380
 in include_from_node1 at .\loading.jl:429
 in macro expansion at .\util.jl:180 [inlined]
 in runtests at C:\projects\julia\test\testdefs.jl:7
 in #16 at C:\projects\julia\test\runtests.jl:44
 in #503 at .\multi.jl:1193
 in run_work_thunk at .\multi.jl:844
 in macro expansion at .\multi.jl:1193 [inlined]
 in #502 at .\event.jl:46
while loading C:\projects\julia\test\strings/basic.jl, in expression starting on line 251
while loading C:\projects\julia\test\string.jl, in expression starting on line 5
```

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17404#issuecomment-232811897) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/06171352e4ea71beabdcc13a4405fa5880aa0b6d/c694fa2_vs_edb112a/report.md). cc @jrevels

**Comment by Sacha0** [comment]
Am I misreading [BaseBenchmarks/SparseBenchmarks.jl](https://github.com/JuliaCI/BaseBenchmarks.jl/blob/master/src/sparse/SparseBenchmarks.jl), or does Nanosoldier not cover sparse `setindex!` performance?

**Comment by tkelman** [comment]
It might not.

File: base/sparse/sparsematrix.jl [hunk 1]
```diff
@@ -2532,23 +2525,59 @@ setindex!(A::SparseMatrixCSC, x, ::Colon, ::Colon) = setindex!(A, x, 1:size(A, 1
...
-
-function spset!{Tv,Ti<:Integer}(A::SparseMatrixCSC{Tv}, x::Tv, I::AbstractVector{Ti}, J::AbstractVector{Ti})
-    !issorted(I) && (I = sort(I))
-    !issorted(J) && (J = sort(J))
-
-    m, n = size(A)
-    lenI = length(I)
-
-    if (!isempty(I) && (I[1] < 1 || I[end] > m)) || (!isempty(J) && (J[1] < 1 || J[end] > n))
+function setindex!{TvA,Tx<:Number,TiI<:Integer,TiJ<:Integer}(A::SparseMatrixCSC{TvA}, x::Tx,
```
Comment (tkelman) [inline base/sparse/sparsematrix.jl]: `Tx` doesn't need to be a type parameter

File: base/sparse/sparsematrix.jl [hunk 2]
```diff
@@ -2532,23 +2525,59 @@ setindex!(A::SparseMatrixCSC, x, ::Colon, ::Colon) = setindex!(A, x, 1:size(A, 1
...
+        I::AbstractVector{TiI}, J::AbstractVector{TiJ})
+    if isempty(I) || isempty(J); return A; end
+    if !issorted(I); I = sort(I); end
+    if !issorted(J); J = sort(J); end
+    if (I[1] < 1 || I[end] > A.m) ||
+         (J[1] < 1 || J[end] > A.n)
```
Comment (tkelman) [inline base/sparse/sparsematrix.jl]: this is short enough that it can probably stay on the same line

File: base/sparse/sparsematrix.jl [hunk 3]
```diff
@@ -2532,23 +2525,59 @@ setindex!(A::SparseMatrixCSC, x, ::Colon, ::Colon) = setindex!(A, x, 1:size(A, 1
...
+function _spsetz_setindex!{TvA,TiI<:Integer,TiJ<:Integer}(A::SparseMatrixCSC{TvA},
+        I::AbstractVector{TiI}, J::AbstractVector{TiJ})
+    lengthI = length(I)
+    for j in J
+        coljAfirstk = A.colptr[j]
+        coljAlastk = A.colptr[j+1] - 1
+        kA = coljAfirstk
+        kI = 1
+        while kA <= coljAlastk && kI <= lengthI
+            # On the one hand, this while structure performs more loads than it should.
```
Comment (tkelman) [inline base/sparse/sparsematrix.jl]: wouldn't be hard to move the load to only happen for the incremented one

File: base/sparse/sparsematrix.jl [hunk 4]
```diff
@@ -2893,13 +2862,17 @@ function setindex!{Tv,Ti}(A::SparseMatrixCSC{Tv,Ti}, x, I::AbstractMatrix{Bool})
                     end
                 end
 
-                # 0: no change, 1: update, 2: delete, 3: add new
-                mode = ((r1 <= r2) && (rowvalA[r1] == row)) ? ((v == 0) ? 2 : 1) : ((v == 0) ? 0 : 3)
+                # 0: no change, 1: update, 3: add new
```
Comment (tkelman) [inline base/sparse/sparsematrix.jl]: no 3 any more

**Comment by tkelman** [comment]
Is there an API (unexported?) for deleting a specific value from the nonzero structure? `dropzeros!` does the whole array, is there something more targeted available if you happen to need that?

File: base/sparse/sparsematrix.jl (hunk 1 above)
Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: Good catch! Related question: To fix the `show` test failure I reverted comparisons of the form `x == zero(Tx)` to the former implementation's `x == 0`, the test failure being due to lack of a `zero` method for some `<:Number`. Is `zero` part of the implicit `Number` interface? That is, in the long run would keeping `x == zero(Tx)` and requiring all `<:Number`s to implement `zero` be better than using `x == 0`, or is `x == 0` the preferred expression? Thanks!

Comment (tkelman) [inline base/sparse/sparsematrix.jl]: I'd rather continue allowing sparse matrices with value types that don't implement `zero` if at all possible

Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: Cheers, the most recent version should make that so. Thanks!

File: base/sparse/sparsematrix.jl (hunk 3 above)
Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: Rewritten avoiding the excess loads. Slightly more complex but still quite readable it seems. Thoughts? Thanks!

File: base/sparse/sparsematrix.jl (hunk 2 above)
Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: Same line now. Thanks!

File: base/sparse/sparsematrix.jl (hunk 4 above)
Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: Mode 3 renumbered 2 in both relevant methods. Those methods could benefit from a rewrite at some point. Thanks!

**Comment by Sacha0** [comment]
> Is there an API (unexported?) for deleting a specific value from the nonzero structure? dropzeros! does the whole array, is there something more targeted available if you happen to need it?

Not at the moment, but we could introduce one now if you would like that facility to appear in the same PR. If so, name? `dropstored!` is the best I've thought of so far. Methods? Perhaps introduce a `dropstored!(A, i, j)` method and rename the existing `spdelete!(A, I, J)` to `dropstored!(A, I, J)` for now? Other forms that should exist immediately? Best!

**Comment by tkelman** [comment]
`dropstored!` does sound like a nicer name than `spdelete!` I suppose. The implementation should be the same or very close, right?

**Comment by Sacha0** [comment]
> dropstored! does sound like a nicer name than spdelete! I suppose. The implementation should be the same or very close, right?

Prototype implementation of `dropstored!` (lacking linear and logical indexing) now included. Thoughts? Thanks!

**Comment by Sacha0** [comment]
The most recent commits should make `SparseVector` behavior consistent with the new `SparseMatrixCSC` behavior. Best!

**Comment by tkelman** [comment]
this may not be all that useful given the likely lack of sparse setindex tests, but it's otherwise probably idle right now so
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17404#issuecomment-233122812) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/edb4d3b7daceffbaf28d8035e5591854944d11a2/4e34433_vs_1fd440e/report.md). cc @jrevels

File: base/sparse/sparsematrix.jl [hunk 5]
```diff
@@ -3072,6 +3044,109 @@ function setindex!{Tv,Ti,T<:Real}(A::SparseMatrixCSC{Tv,Ti}, x, I::AbstractVecto
...
+                    end
+                    rowidx += 1
+                end
+            end
+        end
+    end
+
+    if ndel > 0
+        A.colptr[n+1] = rowidx
+        deleteat!(rowvalA, rowidx:nnzA)
```
Comment (KristofferC) [inline base/sparse/sparsematrix.jl]: This could use `resize!` instead?

File: base/sparse/sparsematrix.jl [hunk 6]
```diff
@@ -3072,6 +3044,109 @@ function setindex!{Tv,Ti,T<:Real}(A::SparseMatrixCSC{Tv,Ti}, x, I::AbstractVecto
...
+function dropstored!{TiI<:Integer,TiJ<:Integer}(A::SparseMatrixCSC,
+        I::AbstractVector{TiI}, J::AbstractVector{TiJ})
+    m, n = size(A)
+    nnzA = nnz(A)
+    (nnzA == 0) && (return A)
+
+    !issorted(I) && (I = sort(I))
+    !issorted(J) && (J = sort(J))
+
+    if (!isempty(I) && (I[1] < 1 || I[end] > m)) || (!isempty(J) && (J[1] < 1 || J[end] > n))
```
Comment (KristofferC) [inline base/sparse/sparsematrix.jl]: Could the bounds checking be rewritten to just use `checkbounds`?

File: base/sparse/sparsematrix.jl [hunk 7]
```diff
@@ -3044,6 +3044,109 @@ function setindex!{Tv,Ti,T<:Real}(A::SparseMatrixCSC{Tv,Ti}, x, I::AbstractVecto
...
+        rrange = nzrange(A, col)
+        if ndel > 0
+            A.colptr[col] = A.colptr[col] - ndel
+        end
+
+        if isempty(rrange) || !(col in J)
+            nincl = length(rrange)
+            if(ndel > 0) && !isempty(rrange)
+                copy!(rowvalA, rowidx, rowval, rrange[1], nincl)
+                copy!(nzvalA, rowidx, nzval, rrange[1], nincl)
```
Comment (tkelman) [inline base/sparse/sparsematrix.jl]: won't these alias if this gets hit before the copy does?

File: base/sparse/sparsematrix.jl [hunk 8]
```diff
@@ -3044,6 +3044,109 @@ function setindex!{Tv,Ti,T<:Real}(A::SparseMatrixCSC{Tv,Ti}, x, I::AbstractVecto
...
+    nzval = nzvalA = A.nzval
+    rowidx = 1
+    ndel = 0
+    @inbounds for col in 1:n
+        rrange = nzrange(A, col)
+        if ndel > 0
+            A.colptr[col] = A.colptr[col] - ndel
+        end
+
+        if isempty(rrange) || !(col in J)
```
Comment (tkelman) [inline base/sparse/sparsematrix.jl]: seems like this (and `rowval[ridx] in I`) should be able to take advantage of `I` and `J` being sorted

**Comment by Sacha0** [comment]
Apologies for the slow response (travel Saturday through Monday).

Re. `dropstored!(A, I, J)`, the implementation is cut-and-paste from the former `spdelete!`. That code would benefit from a thorough rewrite. And likewise for the similarly-written `setindex!` code. Which is to say that I wholeheartedly agree with the corresponding comments. I forwent most such rewriting to minimize replacement of well-exercised code with unexercised code just prior to the anticipated release candidate freeze and alongside related breaking changes. If such rewriting would be better in this rather than a later PR, however, I can take a shot at it now (timescale unclear). Thanks and best!

**Comment by tkelman** [comment]
I think we'd be fine to merge this now and do cleanup later. Couldn't tell from the diffs that this was reused code.

**Comment by Sacha0** [comment]
> Couldn't tell from the diffs that this was reused code.

Sorry about that. Reshuffling the contents of the commits such that `spdelete!` clearly transforms into `dropstored!` might be possible. Should I attempt that?

**Comment by tkelman** [comment]
No, I don't think that's worth the hassle right now. There are other more pressing issues that could use your input.

**Comment by KristofferC** [comment]
+1 for merge as is

**Comment by Sacha0** [comment]
Thanks for the reviews and merge!

File: base/sparse/sparsematrix.jl [hunk 9]
```diff
@@ -3072,6 +3044,109 @@ function setindex!{Tv,Ti,T<:Real}(A::SparseMatrixCSC{Tv,Ti}, x, I::AbstractVecto
...
+        deleteat!(A.rowval, searchk)
+        deleteat!(A.nzval, searchk)
+        @simd for m in j:(A.n + 1)
+            @inbounds A.colptr[m] -= 1
+        end
+    end
+    return A
+end
+"""
+    dropstored!(A::SparseMatrix, I::AbstractVector{<:Integer}, J::AbstractVector{<:Integer})
```
Comment (tkelman) [inline base/sparse/sparsematrix.jl]: should this say SparseMatrixCSC ?

Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: #20664. Thanks!

**Pull request merged**